MediaRecorder.stop() fires an additional dataavailable event with bytes after MediaRecorder.pause()

youennf opened the pull request:
#### f2967879748b0e31bc39b20034bb593097555a64
<pre>
MediaRecorder.stop() fires an additional dataavailable event with bytes after MediaRecorder.pause()
<a href="https://bugs.webkit.org/show_bug.cgi?id=243837">https://bugs.webkit.org/show_bug.cgi?id=243837</a>
rdar://problem/98565949

Reviewed by Eric Carlson.

When fetching recorded data, make sure to flush encoders to get all possible data.
This allows to ensure that we get as much data as possible when requestData is called.
This aligns with other browsers behavior.

Covered by added test.

* LayoutTests/http/wpt/mediarecorder/pause-recording-expected.txt:
* LayoutTests/http/wpt/mediarecorder/pause-recording.html:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
(WebCore::AudioSampleBufferCompressor::finish):
(WebCore::AudioSampleBufferCompressor::flush):
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::flushInternal):
(WebCore::AudioSampleBufferCompressor::finish): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(WebCore::MediaRecorderPrivateWriter::fetchData):
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.h:
(WebCore::VideoSampleBufferCompressor::finish):
(WebCore::VideoSampleBufferCompressor::flush):
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.mm:
(WebCore::VideoSampleBufferCompressor::flushInternal):
(WebCore::VideoSampleBufferCompressor::finish): Deleted.

Canonical link: <a href="https://commits.webkit.org/253529@main">https://commits.webkit.org/253529@main</a>
</pre>

eric-carlson commented on Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
```diff
@@ -485,17 +485,25 @@ static inline void appendEndsPreviousSampleDurationMarker(AVAssetWriterInput *as
-        callOnMainThread([weakThis = WTFMove(weakThis)] {
-            if (weakThis)
-                weakThis->completeFetchData();
+            callOnMainThread([weakThis = WTFMove(weakThis)] {
+                if (weakThis)
+                    weakThis->completeFetchData();
+            });
```
Is this `callOnMainThread` necessary, aren't we are already on the main thread?

youennf replied:
It is preexisting but you might be right this is not strictly needed.
I'll check whether we can remove it in a follow-up.

eric-carlson reviewed (approved):
r=me

youennf committed:
MediaRecorder.stop() fires an additional dataavailable event with bytes after MediaRecorder.pause()
https://bugs.webkit.org/show_bug.cgi?id=243837
rdar://problem/98565949

Reviewed by Eric Carlson.

When fetching recorded data, make sure to flush encoders to get all possible data.
This allows to ensure that we get as much data as possible when requestData is called.
This aligns with other browsers behavior.

Covered by added test.

* LayoutTests/http/wpt/mediarecorder/pause-recording-expected.txt:
* LayoutTests/http/wpt/mediarecorder/pause-recording.html:
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.h:
(WebCore::AudioSampleBufferCompressor::finish):
(WebCore::AudioSampleBufferCompressor::flush):
* Source/WebCore/platform/mediarecorder/cocoa/AudioSampleBufferCompressor.mm:
(WebCore::AudioSampleBufferCompressor::flushInternal):
(WebCore::AudioSampleBufferCompressor::finish): Deleted.
* Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterCocoa.mm:
(WebCore::MediaRecorderPrivateWriter::fetchData):
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.h:
(WebCore::VideoSampleBufferCompressor::finish):
(WebCore::VideoSampleBufferCompressor::flush):
* Source/WebCore/platform/mediarecorder/cocoa/VideoSampleBufferCompressor.mm:
(WebCore::VideoSampleBufferCompressor::flushInternal):
(WebCore::VideoSampleBufferCompressor::finish): Deleted.

Canonical link: https://commits.webkit.org/253529@main
